### ISSUE TITLE
Retry Policy for Retrofit

### ISSUE THREAD

**Issue by rt4914** [oppia/oppia-android#76]
The data module is responsible for all the API calls and there should be a retry policy when API call fails. Retrofit does not have retry policy by default. We should probably use OkHttp interceptor to solve this issue. We will be handling standard error codes centrally and the error codes 408-Request Timeout, 502-Bad Gateway, 503-Service Unavailable and 504-Gateway Timeout will result in exponential backoff policy with limited number of retries.
For example, we can have 3 retries (with 2, 4 and 10-second waits between successive retries). The exponential  policy with limited retires will avoid overloading the Oppia backend and will also use limited mobile data of the user.

This is blocked on #5 

----------------------------------------
### Context
Currently, Retrofit/OkHttp does not provide built-in retry policies. To make the app more resilient, we need a retry mechanism for transient HTTP errors. In the current implementation, Retrofit is set up in [RetrofitModule](https://github.com/oppia/oppia-android/blob/96af5e5f11dbf856eef5be9f0ccf264fde4491d6/data/src/main/java/org/oppia/android/data/backends/gae/RetrofitModule.kt#L14) using OkHttpClient, but no retry logic is present. A clean solution can be integrated by introducing a custom `RetryInterceptor` without affecting the existing interceptors for authentication, logging, and JSON prefix handling.

**The target behavior is:**
- Retry for HTTP status codes: 408 (Request Timeout), 502 (Bad Gateway), 503 (Service Unavailable), 504 (Gateway Timeout)
- Retry up to 3 times
- Use exponential backoff delays of 2s → 4s → 10s
- If retries are exhausted, fall back to the original failure.

### Implementation Plan

1. **Add a new file:** `RetryInterceptor` under `data/src/main/java/org/oppia/android/data/backends/gae
`

   **The interceptor should**:
    - Intercept all outgoing requests.
    - Retry for error codes 408, 502, 503, 504.
    - Apply exponential backoff with delays 2s → 4s → 10s.
    - Stop after 3 attempts and return the final failure.

     👉 For reference, see this [article on implementing retry logic with OkHttp interceptors](https://medium.com/mobilepeople/how-to-retry-network-requests-automatically-in-android-kotlin-64dcafb7f294)

2. **Update [RetrofitModule](https://github.com/oppia/oppia-android/blob/96af5e5f11dbf856eef5be9f0ccf264fde4491d6/data/src/main/java/org/oppia/android/data/backends/gae/RetrofitModule.kt#L14)**
    - Add the new `RetryInterceptor` to the [OkHttpClient.Builder](https://github.com/oppia/oppia-android/blob/96af5e5f11dbf856eef5be9f0ccf264fde4491d6/data/src/main/java/org/oppia/android/data/backends/gae/RetrofitModule.kt#L29).
    - Place it after the auth interceptor but before logging and JSON prefix interceptors. (This ensures retries happen after authentication but before logging and response modification.)

3. **Add a new file:** `RetryInterceptorTest` under `data/src/test/java/org/oppia/android/data/backends/gae/testing` to verify the retry logic.

**Comment by Mehul-Bisht** [comment]
Is this issue still blocked?

**Comment by subhajitxyz** [comment]
**The original issue proposed implementing a retry policy in the data module using an OkHttp interceptor to handle transient HTTP errors (408, 502, 503, 504) with exponential backoff (2s, 4s, 10s) for up to 3 retries.**

After reviewing the current Retrofit implementation, it's confirmed that:

- Retrofit setup exists in RetrofitModule using OkHttpClient
- No retry logic is currently implemented
- A clean solution using a custom `RetryInterceptor` can be integrated without affecting existing interceptors (auth, logging, JSON prefix)


**Implement a retry policy using an OkHttp interceptor (RetryInterceptor) that:**

- Retries failed requests for specific transient HTTP status codes: 408 (Request Timeout), 502 (Bad Gateway), 503 (Service Unavailable), 504 (Gateway Timeout)
- Uses an exponential backoff strategy with delays of 2s → 4s → 10s
- Retries up to 3 times
- Falls back to the original failure if retries are exhausted

**Implementation Plan**
For reference, see this [article on implementing retry logic with OkHttp interceptors](https://medium.com/mobilepeople/how-to-retry-network-requests-automatically-in-android-kotlin-64dcafb7f294)

- Add new file RetryInterceptor.kt in: `data/src/main/java/org/oppia/android/data/backends/gae`

- Integration in existing RetrofitModule: Update the OkHttpClient.Builder in RetrofitModule to include the retry interceptor:

**In RetrofitModule**
```
return Retrofit.Builder().apply {
      baseUrl(baseUrl)
      addConverterFactory(MoshiConverterFactory.create(moshi))
      client(
        OkHttpClient.Builder().apply {
          // This is in a specific order. The auth modifies a request, so it happens first. The
          // prefix remover executes other interceptors before changing the response, so it's
          // registered last so that the network logging interceptor receives a response with the
          // XSSI prefix correctly removed.
          addInterceptor(remoteAuthNetworkInterceptor)
          addInterceptor(retryInterceptor)
          addInterceptor(networkLoggingInterceptor)
          addInterceptor(jsonPrefixNetworkInterceptor)
        }.build()
      )
    }.build()
```

**Comment by subhajitxyz** [comment]
Hi @adhiamboperes  , PTAL at this finding for GFI issue breakdown and share your feedback. Then I will add it in issue description.

**Comment by adhiamboperes** [comment]
@subhajitxyz, this description does not clearly point towards an expected solution. It would be helpful to maybe link an SO or blog link that can be helpful to new contributors to handle the retries.

**Comment by subhajitxyz** [comment]
@adhiamboperes , I have updated the breakdown, PTAL.

**Comment by adhiamboperes** [comment]
@subhajitxyz, please add the solution to the description, but with the following changes:
1.  Mention that the `RetryInterceptor` is a new file and the directory or package where it should be added.
2. Remove the code example of the RetryInterceptor. The contributor should try to write their own code so that we can see how they work. You can keep the `RetrofitModule` usage example however.

**Comment by subhajitxyz** [comment]
Fixed.

**Comment by omkodes** [comment]
Hi team! 

I would love to work on this issue.

I've started planning the implementation, and my proposed approach is as follows:

### **Implementation Plan**

**1. Create a `RetryInterceptor`**
- Implement a custom OkHttp `Interceptor` that retries requests for **408, 502, 503, and 504** status codes.

**2. Implement Exponential Backoff**
- Retry failed requests up to **3 times** with delays of **2s, 4s, and 10s**, ensuring `response.close()` is called before each retry to prevent resource leaks.

**3. Dependency Injection Integration**
- Register the interceptor in `RetrofitModule.kt` and add it to the `OkHttpClient.Builder` after `RemoteAuthNetworkInterceptor` and before the logging and JSON prefix interceptors.

**4. Testing**
- Write comprehensive unit tests in a new `RetryInterceptorTest` class to validate the retry interceptor implementation.

I'd be happy to work on this issue. Could you please assign it to me?

Thank you!

**Comment by Neer-rn** [comment]
Hi @omkodes, thanks for the approached plan, it matches with the issue breakdown also. Assigning this to you.

**Comment by adhiamboperes** [comment]
@omcodes, would you be interested in finishing the leftover todos that caused the issue to reopen?

**Comment by omkodes** [comment]
Hi @adhiamboperes! Yes, I'd love to work on this.

Before proceeding, could you clarify does the Oppia backend return any JSON body when `postFeedbackReport` fails? If yes, what does that error model look like? That'll help me decide whether to model a proper error payload or just remove the TODO comments.

**Comment by adhiamboperes** [comment]
Hi @omkodes, apologies, I took a look at the todos and it does not look like they can be addressed at this point. For example, the feedback service hasn't been fully implemented yet. We will leave this issue open for now and remove the good first issue label. I am sending you another task over email.

**Comment by omkodes** [comment]
Thank you for the clarification, @adhiamboperes! That makes complete sense. I'll leave those TODOs as-is since the feedback service hasn't been fully implemented yet. I've already received the new task over email and will get started on it.
